### PR TITLE
Feat: ScrollScreenshot

### PR DESCRIPTION
**Pull request by zbysir**
Add ScrollScreenshot, for https://github.com/go-rod/rod/issues/1008

### CODE REVIEW

File: lib/examples/compare-chromedp/scroll_screenshot/main.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// Package main ...
+package main
```
Comment (ysmood) [inline lib/examples/compare-chromedp/scroll_screenshot/main.go:2]: Could you move this to this file: https://github.com/go-rod/rod/blob/main/examples_test.go

File: page.go [hunk 2]
```diff
@@ -464,6 +464,126 @@ func (p *Page) Screenshot(fullPage bool, req *proto.PageCaptureScreenshot) ([]by
...
+
+	// WaitPreScroll wait scroll animation (default is 500ms)
+	WaitPreScroll time.Duration
+}
+
+// ScrollScreenshot Scroll screenshot does not adjust the size of the viewport, but achieves it by scrolling and capturing screenshots in a loop, and then stitching them together.
+// Note that this method also has a flaw: when there are elements with fixed positioning on the page (usually header navigation components), these elements will appear repeatedly,
+//
+//	you can set the FixedTop parameter to optimize it.
+func (p *Page) ScrollScreenshot(opt *ScrollScreenshotOptions) ([]byte, error) {
```
Comment (ysmood) [inline page.go:489]: Could you add a test case to `page_test.go`, or the CI will fail, we require 100% code coverage.

Check this doc: https://github.com/go-rod/rod/blob/main/.github/CONTRIBUTING.md#run-tests

**Review by ysmood — CHANGES_REQUESTED** [review]

**Comment by ysmood** [comment]
Good job! It will be great if you could add some tests.

File: lib/examples/compare-chromedp/scroll_screenshot/main.go (hunk 1 above)
Comment (zbysir) [reply lib/examples/compare-chromedp/scroll_screenshot/main.go:2]: I'm finished. Any other questions?

**Comment by zbysir** [comment]
> Good job! It will be great if you could add some tests.

Review again, please allow me to make some weird mistakes.

**Review by ysmood — CHANGES_REQUESTED** [review]
Please fix the lint error in CI, you can check it on your local first `go run ./lib/utils/lint`.

The test doesn't work on Mac and Windows: https://github.com/go-rod/rod/actions/runs/7751056459/job/21138412821?pr=1010#step:4:25

**Comment by zbysir** [comment]
> Please fix the lint error in CI, you can check it on your local first `go run ./lib/utils/lint`.
> 
> The test doesn't work on Mac and Windows: https://github.com/go-rod/rod/actions/runs/7751056459/job/21138412821?pr=1010#step:4:25

Yes, I fixed it. Done.

**Comment by ysmood** [comment]
Please check the CI, you need help let me know.

**Comment by zbysir** [comment]
@ysmood  Need help:

The test code cannot cover the error branch below

```go
              err = p.WaitStable(opt.WaitPreScroll)
                if err != nil {
                        return nil, fmt.Errorf("waitStable error: %w", err)
                }
```
I don't know how to fix it, I tried the test code below but it doesn't work
```go
              err = p.WaitStable(opt.WaitPreScroll)
                if err != nil {
                        return nil, fmt.Errorf("waitStable error: %w", err)
                }
```

Do you have any plans? Or is it appropriate for me to use time.Sleep() directly?

**Review by zbysir — COMMENTED** [review]
`go run ./lib/utils/lint` will end normally now

**Review by zbysir — COMMENTED** [review]
Finished

File: page.go [hunk 3]
```diff
@@ -464,6 +464,126 @@ func (p *Page) Screenshot(fullPage bool, req *proto.PageCaptureScreenshot) ([]by
...
+		if scrollTop >= contentHeight {
+			break
+		}
+		err = p.Mouse.Scroll(0, scrollY, 1)
+		if err != nil {
+			return nil, fmt.Errorf("scroll error: %w", err)
+		}
+
+		err = p.WaitStable(opt.WaitPreScroll)
+		if err != nil {
```
Comment (zbysir) [inline page.go:570]: I can't cover this error branch in my tests, please help, bro.

**Review by zbysir — COMMENTED** [review]
Updated

File: page.go (hunk 3 above)
Comment (ysmood) [reply page.go:570]: CleanShot 2024-02-02 at 19 41 23@2x

Comment (zbysir) [reply page.go:570]: I tried, mentioned in the comments above.

Comment (ysmood) [reply page.go:570]: Use `g.mc.stubErr(1, proto.DOMGetContentQuads{})`

Comment (ysmood) [reply page.go:570]: Your code uses `DOMSnapshotCaptureSnapshot`, but we don't want this cdp call to fail right? You want some cdp call in WaitStable to fail, `DOMGetContentQuads` is one that WaitStable uses.

Comment (zbysir) [reply page.go:570]: > DOMGetContentQuads

It seems like there are still some problems:


```
	g.Panic(func() {
		// mock error for WaitStable
		g.mc.stubErr(1, proto.DOMGetContentQuads{})
		p.MustScrollScreenshotPage()
	})
```

--- FAIL: TestScrollScreenshotPage (47.14s)
    page_test.go:805:  ⦗should panic⦘

Comment (zbysir) [reply page.go:570]: > Your code uses `DOMSnapshotCaptureSnapshot`, but we don't want this cdp call to fail right? You want some cdp call in WaitStable to fail, `DOMGetContentQuads` is one that WaitStable uses.

Haha, I saw that `WaitDOMStable` would be called `DOMSnapshotCaptureSnapshot` req, so I gave it a try.

Comment (ysmood) [reply page.go:570]: Got it, let me run the code

Comment (ysmood) [reply page.go:570]: I pushed a race condition fix on the main branch: https://github.com/go-rod/rod/commit/4d3ff679eaa440737b2dab06f2f4660f83dc6b92

Then this works for me:

```go
	g.Panic(func() {
		// mock error for Screenshot
		g.mc.stubErr(1, proto.DOMSnapshotCaptureSnapshot{})
		p.MustScrollScreenshotPage()
	})
```

Comment (zbysir) [reply page.go:570]: > I pushed a race condition fix on the main branch: [ce2a1af](https://github.com/go-rod/rod/commit/ce2a1afc29647fd9ca403829333a733fd39bffc9)
> 
> Then this works for me:
> 
> ```go
> 	g.Panic(func() {
> 		// mock error for Screenshot
> 		g.mc.stubErr(1, proto.DOMSnapshotCaptureSnapshot{})
> 		p.MustScrollScreenshotPage()
> 	})
> ```

Yes, I resubmitted the code. Please try running CI again. It’s just a little bit off.

**Comment by zbysir** [comment]
@ysmood Please run CI again and it will almost succeed.

**Review by ysmood — APPROVED** [review]
👍🏼

**Comment by zbysir** [comment]
@ysmood Thank you for your patient help!

**Pull request merged**